Introduce scope tracking in the parser

danez opened the pull request:
| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            |  Fixes #6722 (Fixes maybe #6283) Fixes #9316 Fixes #4331
| Patch: Bug Fix?          | y
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT (acorn MIT)

This introduces scope tracking in the parser. The scope tracking replaces some of the Flags we currently have (inGenerator, inFunction, ...).

Spec changes:

- We now report an syntax error if `super` is used even though there is no superClass
- Duplicate bindings will throw according to spec

The idea here is to enter and exit scopes, during parsing so we always can check the current scope. A scope consists of flags (which adds information we need in several places like SCOPE_FUNCTION, SCOPE_ASYNC, ...) and 3 collections of declared bindings in this scope (var, lexical(let, const, classes) and functions).

TODO:

- [x] Add tests for duplicate variable bindings.
- [x] Migrate some traverse tests (duplicate declarations) to not use the parser

Big thanks to acorn, as most of this here is ported from acorn. :) Same goes for copyright on most of it.

danez commented:
We have some tests that test the duplicate declaration checks of babel-traverse, but as the parser now throws first we are not getting to the point of testing duplicate declarations in traverse.

I see 3 options:

1. Leave the tests as is and don't care that we are not testing traverse.
2. Add an option to the parser to disable the duplicate check.
3. Move the tests to the parser (as thats what they are testing now) but also recreate new tests for traverse which do not use the parser but node builders instead to create the AST.


1 and 2 are not really good options for me. So I guess I will go for 3 which is the most work but the best I fell? Any other ideas on this?

nicolo-ribaudo commented:
:+1: for 3. We could ask help to the community to rewrite the `@babel/traverse` tests.

danez commented:
> 👍 for 3. We could ask help to the community to rewrite the `@babel/traverse` tests.

Turns out this are only 4 tests. Probably took longer to write the message above, than to fix them :D

danez commented on packages/babel-plugin-transform-classes/src/transformClass.js:
```diff
@@ -19,22 +19,6 @@ function buildConstructor(classRef, constructorBody, node) {
-    },
-  },
-]);
-
```
Not sure if i should leave that or not. If this check would be in traverse I would leave it, but in the transform, the only valid usecase now would be someone creating in a plugin invalid code and then running the transform.

nicolo-ribaudo replied:
I don't think that we need it: currently all the other early errors (except for the duplicated bindings errors) are only in `@babel/parser`.

danez commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -564,6 +576,7 @@ export default class StatementParser extends ExpressionParser {
     const cases = (node.cases = []);
     this.expect(tt.braceL);
     this.state.labels.push(switchLabel);
+    this.enterScope(0);
```
makes sense

danez commented on packages/babel-parser/src/parser/scope.js:
```diff
@@ -0,0 +1,152 @@
+
+// The functions in this module keep track of declared variables in the
+// current scope in order to detect duplicate variable names.
+export default class ScopeParser extends UtilParser {
```
I changed it to be a separate class, which works and might be nice to have like this if we would want to reuse this parts in traverse for example.

But when looking into flow I noticed another problem: extending the scope tracking is not as easy possible with the external class. So for typescript we probably would want to add another binding group for types. Flow has some custom rules too when it comes to `declare class|var|function`. I'm not sure if we actually want to support extending the scope tracking but it might be nice.

danez commented:
Okay so I added fixes for flow:
- declare module {} has it's own scope now
- `type` and `opaque type` are now registered as lexical declaration, so the are handled the same as `let` or `const`

I also added test for TS to ensure types do not count as duplicate declarations of other lexical declarations.

Any further additions for flow and typescript need changes to the logic in how duplicate declarations are checked. But the problem is that there is no easy way to extend the scope handler, as any addition (for example a new scope group for typescript type declarations) will need adjustments to all other parts of the scope tracking logic.
The only way to achieve correct scope tracking for flow and ts is to check on initialization which language we are parsing and depending on that choose a corresponding ScopeHandler. This ScopeHandlers for JS, TS and flow ultimately will need to be copy and pasted and kept in sync for bug fixes as I do not see a way to easy overwrite a subset of the logic of the duplicate detection algorithm.

For now this PR does not introduce any duplicate checks for flow an ts, besides the ones mentions above, and we can follow up in a separate PR.

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -954,59 +989,66 @@ export default class StatementParser extends ExpressionParser {
     this.initFunction(node, isAsync);
 
+    if (this.match(tt.star) && isHangingStatement) {
+      this.unexpected();
```
Maybe we can have a better error message here?

Chrome throws these errors:
```
Generators can only be declared at the top level or inside a block.

Async functions can only be declared at the top level or inside a block.
```

danez replied:
I updated them and some more error messages

nicolo-ribaudo commented on packages/babel-parser/src/parser/util.js:
```diff
@@ -3,7 +3,9 @@
-import { lineBreak } from "../util/whitespace";
+import { lineBreak, skipWhiteSpace } from "../util/whitespace";
+
+const literal = /^(?:'((?:\\.|[^'])*?)'|"((?:\\.|[^"])*?)"|;)/;
```
I took 5 mins to understand what is this lol

I think that it could be simplified a bit:
```js
/^(?:;|('|")((?:\\?.)*?)\1)/
```

1. You don't need the `[^"]` part, because you are already using `*?` so it stops as soon as possible (and the closing quote this is parsed as the end of the literal).
2. You can merge the `"` and `'` cases, using `\1` to refer to the matched quotes kind.
3. Since this adds a new capturing group, you would have to chech `match[2]`.

Demo: https://regex101.com/r/aQTDzs/1

danez replied:
Thanks, that is way easier to read.
